File: bindings/dart/benchmark/benchmark.dart
// benchmark/benchmark.dart
import 'dart:io';
import 'package:wavedb/wavedb.dart';

const int iterations = 10000;
const int warmup = 100;

void cleanPath(String dbPath) {
  final dir = Directory(dbPath);
  if (dir.existsSync()) dir.deleteSync(recursive: true);
}

Future<void> warmupDb(WaveDB db) async {
  for (var i = 0; i < warmup; i++) {
    await db.put('warmup$i', 'value$i');
    await db.get('warmup$i');
  }
}

Future<void> benchSequentialAsync(WaveDB db, String label) async {
  print('\nSequential Async ($label):');
  print('-' * 50);

  var sw = Stopwatch()..start();
  for (var i = 0; i < iterations; i++) {
    await db.put('key$i', 'value$i');
  }
  sw.stop();
  var ops = (iterations / sw.elapsedMilliseconds * 1000).round();
  print('  put:       ${ops.toString().padLeft(8)} ops/sec');

  sw = Stopwatch()..start();
  for (var i = 0; i < iterations; i++) {
    await db.get('key$i');
  }
  sw.stop();
  ops = (iterations / sw.elapsedMilliseconds * 1000).round();
  print('  get:       ${ops.toString().padLeft(8)} ops/sec');
}

Future<void> benchSync(WaveDB db, String label) async {
  print('\nSync Operations ($label):');
  print('-' * 50);

  var sw = Stopwatch()..start();
  for (var i = 0; i < iterations; i++) {
    db.putSync('sync$i', 'value$i');
  }
  sw.stop();
  var ops = (iterations / sw.elapsedMilliseconds * 1000).round();
  print('  putSync:   ${ops.toString().padLeft(8)} ops/sec');

  sw = Stopwatch()..start();
  for (var i = 0; i < iterations; i++) {
    db.getSync('sync$i');
  }
  sw.stop();
  ops = (iterations / sw.elapsedMilliseconds * 1000).round();
  print('  getSync:   ${ops.toString().padLeft(8)} ops/sec');
}

Future<void> benchBatch(WaveDB db) async {
  print('\nBatch Operations:');
  print('-' * 50);
  const batchSize = 1000;
  final batchOps = List.generate(
      batchSize, (i) => {'type': 'put', 'key': 'batch$i', 'value': 'value$i'});
  final batchCount = iterations ~/ batchSize;

  final sw = Stopwatch()..start();
  for (var i = 0; i < batchCount; i++) {
    await db.batch(batchOps);
  }
  sw.stop();
  final ops = (iterations / sw.elapsedMilliseconds * 1000).round();
  print('  batch:    ${ops.toString().padLeft(8)} ops/sec ($batchSize ops/batch)');
}

Future<void> benchConcurrent(WaveDB db, String label) async {
  print('\nConcurrent Async ($label):');
  print('-' * 50);

  final cpuCount = Platform.numberOfProcessors;
  final concurrencies = [1, 2, 4, 8, 16, 32].where((c) => c <= cpuCount * 2).toList();
  if (!concurrencies.contains(1)) concurrencies.insert(0, 1);

  for (final c in concurrencies) {
    // Concurrent puts
    final putKeys = List.generate(iterations, (i) => 'cp${c}_$i');
    var sw = Stopwatch()..start();
    var idx = 0;
    while (idx < iterations) {
      final chunk = putKeys.sublist(idx, (idx + c).clamp(0, iterations));
      await Future.wait(chunk.map((k) => db.put(k, 'v$idx')));
      idx += c;
    }
    sw.stop();
    var ops = (iterations / sw.elapsedMilliseconds * 1000).round();
    print('  put (c=${c.toString().padLeft(2)}):  ${ops.toString().padLeft(8)} ops/sec');

    // Concurrent gets
    sw = Stopwatch()..start();
    idx = 0;
    while (idx < iterations) {
      final chunk = putKeys.sublist(idx, (idx + c).clamp(0, iterations));
      await Future.wait(chunk.map((k) => db.get(k)));
      idx += c;
    }
    sw.stop();
    ops = (iterations / sw.elapsedMilliseconds * 1000).round();
    print('  get (c=${c.toString().padLeft(2)}):  ${ops.toString().padLeft(8)} ops/sec');
  }
}

Future<void> benchStream(WaveDB db) async {
  print('\nStream Operations:');
  print('-' * 50);
  final sw = Stopwatch()..start();
  var count = 0;
  try {
    await for (final _ in db.createReadStream()) {
      count++;
    }
  } catch (_) {}
  sw.stop();
  final ops = count > 0 ? (count / sw.elapsedMilliseconds * 1000).round() : 0;
  print('  stream:    ${ops.toString().padLeft(8)} entries/sec ($count entries)');
}

Future<void> runBenchmarks(
    String dbPath, WaveDBConfig? config, String label) async {
  cleanPath(dbPath);
  final db = WaveDB(dbPath, config: config);
  await warmupDb(db);

  await benchSequentialAsync(db, label);
  await benchSync(db, label);
  await benchBatch(db);
  await benchConcurrent(db, label);
  await benchStream(db);

  db.close();
  cleanPath(dbPath);
}

Future<void> main() async {
  print('WaveDB Dart FFI Benchmarks');
  print('=' * 50);
  print('Dart ${Platform.version}');
  print('Iterations: $iterations');
  print('CPUs: ${Platform.numberOfProcessors}');

  // ---- In-memory (no persistence) ----
  print('\n' + '=' * 50);
  print('MODE: In-Memory (enablePersist=false)');
  print('=' * 50);
  await runBenchmarks('/tmp/wavedb_dart_bench_mem', const WaveDBConfig(enablePersist: false, lruMemoryMb: 50), 'in-memory');

  // ---- DEBOUNCED WAL ----
  print('\n' + '=' * 50);
  print('MODE: DEBOUNCED WAL (100ms fsync)');
  print('=' * 50);
  await runBenchmarks('/tmp/wavedb_dart_bench_debounced', const WaveDBConfig(walSyncMode: 'debounced', walDebounceMs: 100, lruMemoryMb: 50), 'DEBOUNCED');

  // ---- ASYNC WAL ----
  print('\n' + '=' * 50);
  print('MODE: ASYNC WAL (OS cache, no fsync)');
  print('=' * 50);
  await runBenchmarks('/tmp/wavedb_dart_bench_async', const WaveDBConfig(walSyncMode: 'async', lruMemoryMb: 50), 'ASYNC');

  // ---- IMMEDIATE WAL ----
  print('\n' + '=' * 50);
  print('MODE: IMMEDIATE WAL (fsync per write)');
  print('=' * 50);
  await runBenchmarks('/tmp/wavedb_dart_bench_immediate', const WaveDBConfig(walSyncMode: 'immediate', lruMemoryMb: 50), 'IMMEDIATE');

  print('\n' + '=' * 50);
}

File: bindings/python/benchmark.py
#!/usr/bin/env python3
"""WaveDB Python binding performance benchmarks.

Mirrors bindings/nodejs/benchmark.js structure: 5 sections × 4 WAL modes.
Reports ops/sec and us/op for cross-binding comparison.

Usage:
    WAVEDB_LIB_PATH=../../build-release/libwavedb.so python benchmark.py
    WAVEDB_LIB_PATH=../../build-release/libwavedb.so python benchmark.py --c-baseline 0.8
"""
from __future__ import annotations

import argparse
import asyncio
import os
import shutil
import sys
import tempfile
import time

from wavedb import WaveDB, WaveDBConfig

ITERATIONS = 5000
WARMUP = 100
BATCH_SIZE = 1000
FANOUTS = [1, 2, 4, 8, 16, 32]


def clean_path(path: str) -> None:
    if os.path.exists(path):
        shutil.rmtree(path)


def perf_ns() -> int:
    return time.perf_counter_ns()


def report(label: str, ops: int, elapsed_ns: int) -> None:
    elapsed_s = elapsed_ns / 1e9
    ops_per_sec = ops / elapsed_s if elapsed_s > 0 else 0
    us_per_op = elapsed_ns / ops / 1000 if ops > 0 else 0
    print(f"  {label:20s} {ops_per_sec:>10.0f} ops/sec  {us_per_op:>8.2f} us/op")


async def warmup_db(db: WaveDB) -> None:
    for i in range(WARMUP):
        await db.put(f"warmup{i}", f"value{i}")
        await db.get(f"warmup{i}")


async def bench_sequential_async(db: WaveDB) -> None:
    print("\nSequential Async:")
    print("-" * 50)
    t0 = perf_ns()
    for i in range(ITERATIONS):
        await db.put(f"key{i}", f"value{i}")
    report("put", ITERATIONS, perf_ns() - t0)

    t0 = perf_ns()
    for i in range(ITERATIONS):
        await db.get(f"key{i}")
    report("get", ITERATIONS, perf_ns() - t0)


def bench_sync(db: WaveDB) -> None:
    print("\nSync Operations:")
    print("-" * 50)
    t0 = perf_ns()
    for i in range(ITERATIONS):
        db.put_sync(f"sync{i}", f"value{i}")
    report("put_sync", ITERATIONS, perf_ns() - t0)

    t0 = perf_ns()
    for i in range(ITERATIONS):
        db.get_sync(f"sync{i}")
    report("get_sync", ITERATIONS, perf_ns() - t0)


async def bench_batch(db: WaveDB) -> None:
    print("\nBatch Operations:")
    print("-" * 50)
    ops = [{"type": "put", "key": f"batch{i}", "value": f"value{i}"} for i in range(BATCH_SIZE)]
    batches = ITERATIONS // BATCH_SIZE
    t0 = perf_ns()
    for _ in range(batches):
        await db.batch(ops)
    report(f"batch ({BATCH_SIZE}/batch)", ITERATIONS, perf_ns() - t0)


async def bench_concurrent(db: WaveDB) -> None:
    print("\nConcurrent Async:")
    print("-" * 50)
    cpu = os.cpu_count() or 4
    fanouts = [f for f in FANOUTS if f <= cpu * 2]
    if 1 not in fanouts:
        fanouts.insert(0, 1)

    for c in fanouts:
        keys = [f"cp{c}_{i}" for i in range(ITERATIONS)]
        # Concurrent puts (chunked by fanout)
        t0 = perf_ns()
        for idx in range(0, ITERATIONS, c):
            chunk = keys[idx:idx + c]
            await asyncio.gather(*[db.put(k, f"v{i}") for i, k in enumerate(chunk)])
        report(f"put (c={c})", ITERATIONS, perf_ns() - t0)

        # Concurrent gets
        t0 = perf_ns()
        for idx in range(0, ITERATIONS, c):
            chunk = keys[idx:idx + c]
            await asyncio.gather(*[db.get(k) for k in chunk])
        report(f"get (c={c})", ITERATIONS, perf_ns() - t0)


def bench_stream(db: WaveDB) -> None:
    print("\nStream Operations:")
    print("-" * 50)
    # NOTE: scan with 50+ keys triggers a pre-existing C-level segfault in
    # bnode_count during trie traversal (not related to path_meta changes).
    # Skip the stream benchmark until that crash is fixed. The segfault
    # can't be caught by Python exception handling.
    print("  stream: SKIPPED (pre-existing scan crash with 50+ keys)")


async def run_benchmarks(db_path: str, config: WaveDBConfig, label: str) -> None:
    clean_path(db_path)
    db = WaveDB(db_path, config=config)
    await warmup_db(db)

    await bench_sequential_async(db)
    bench_sync(db)
    await bench_batch(db)
    await bench_concurrent(db)
    bench_stream(db)

    await db.aclose()
    clean_path(db_path)


async def main() -> None:
    parser = argparse.ArgumentParser(description="WaveDB Python benchmarks")
    parser.add_argument("--c-baseline", type=float, default=None,
                        help="C microbench us/op for overhead factor comparison")
    args = parser.parse_args()

    print("WaveDB Performance Benchmarks (Python)")
    print("=" * 50)
    print(f"Python {sys.version.split()[0]}")
    print(f"Iterations: {ITERATIONS}")
    print(f"CPUs: {os.cpu_count()}")

    base = tempfile.mkdtemp(prefix="wavedb_bench_")

    # In-memory — use wal_sync_mode="none" (OS cache, no fsync) instead of
    # in_memory=True because async ops + in_memory=True is not yet supported
    # (C worker threads try to access WAL which doesn't exist in in-memory mode).
    # wal_sync_mode="none" gives close-to-in-memory performance with WAL enabled.
    print(f"\n{'=' * 50}")
    print("MODE: In-Memory (wal_sync_mode=none, OS cache)")
    print("=" * 50)
    await run_benchmarks(f"{base}/mem", WaveDBConfig(wal_sync_mode="none"), "in-memory")

    # Debounced WAL
    print(f"\n{'=' * 50}")
    print("MODE: DEBOUNCED WAL (100ms)")
    print("=" * 50)
    await run_benchmarks(
        f"{base}/debounced",
        WaveDBConfig(wal_sync_mode="debounced", wal_debounce_ms=100),
        "DEBOUNCED",
    )

    # Async WAL (OS cache, no fsync)
    print(f"\n{'=' * 50}")
    print("MODE: ASYNC WAL (OS cache, no fsync)")
    print("=" * 50)
    await run_benchmarks(
        f"{base}/async",
        WaveDBConfig(wal_sync_mode="none"),
        "ASYNC",
    )

    # Immediate WAL (fsync per write)
    print(f"\n{'=' * 50}")
    print("MODE: IMMEDIATE WAL (fsync per write)")
    print("=" * 50)
    await run_benchmarks(
        f"{base}/immediate",
        WaveDBConfig(wal_sync_mode="immediate"),
        "IMMEDIATE",
    )

    if args.c_baseline:
        print(f"\n{'=' * 50}")
        print(f"C baseline: {args.c_baseline} us/op")
        print("(Compare against sync put_sync us/op above for overhead factor)")

    shutil.rmtree(base)
    print(f"\n{'=' * 50}")
    print("Done.")


if __name__ == "__main__":
    asyncio.run(main())